qutip/core/data/csc.pyx

```diff
@@ -395,7 +399,7 @@ cpdef CSC from_tr_csr(CSR matrix, bint copy=True):
     If copy is False, steal data ownership from the CSR.
     """
     if not matrix._deallocate and not copy:
-        raise ValueError("Input matrix should own its data.")
+        raise ValueError("Input matrix should own its data if not copying.")
```

Comment:
This now fails in the case that the user passes in a matrix which is already scipy-backed, which prevents people from using this function in Python-space code.  We'd designed the data layer with the idea that you can extend our builtin specialities from Python space,  but since a lot of matrices creating in Python space _will_ be scipy-backed, this generally breaks that.

If you always create a scipy view onto the base matrix, you can safely do this without copying by storing scipy views onto the same numpy arrays, and that avoids the need to reinvent refcounting elsewhere.  I don't think the data layer should be raising exceptions for anything other than a completely impossible user error, and from Python space (which this function is accessible to) a developer should never have to know whether a C object owns its data or not.